src/cpp/src/lora_adapter.cpp

```diff
@@ -0,0 +1,1296 @@
+            ov::Tensor(lora_var_ids.B.data_type, dynamic_to_static(lora_var_ids.B.data_shape))
+        };
+        auto new_tensors = prepare_lora_tensors(lora_indices.name, weight_getters, lora_state_tensors);
+
+        state[lora_indices.alpha].set_state(new_tensors.alpha);
+        state[lora_indices.A].set_state(new_tensors.A);
+        state[lora_indices.B].set_state(new_tensors.B);
+    }
+
+    LoRAParts<ov::Tensor> prepare_lora_tensors (
+        const std::string& name,
+        const std::vector<LoRAWeightGetter>& weight_getters,
+        LoRAParts<ov::Tensor>& output,
+        bool set_empty_adapters = true
+    ) {
+        auto lora_tensors = collect_applicable_tensors(name, weight_getters);
+        LoRAParts<ov::Tensor> new_tensors;
+        if(!lora_tensors.empty()) {
+            new_tensors = concat_adapters(lora_tensors, output);
+        } else if(set_empty_adapters) {  // FIXME: Make it as a separate step outside of this function
+            new_tensors = empty_adapters(lora_tensors, output);
+        }
+        return new_tensors;
+    }
+
+    void apply (ov::InferRequest& infer_request) {
+        return apply(infer_request, current_config);
+    }
+};
+
+
+AdapterController::AdapterController(std::shared_ptr<ov::Model> model, const AdapterConfig& config, const std::string& prefix, std::string device)
+{
+    // If AdapterConfig::MODE_AUTO is used, then set real mode depending on the device capabilities
+    // TODO: Remove this code when devices become aligned on their capabilities for LoRA adapaters
+    if (config.get_mode() == AdapterConfig::MODE_AUTO) {
+        static const std::map<std::string, AdapterConfig::Mode> default_modes {
+            {"CPU", AdapterConfig::MODE_DYNAMIC},
+            {"GPU", AdapterConfig::MODE_STATIC_RANK},
+            {"NPU", AdapterConfig::MODE_STATIC},
+        };
+        if(device.find("GPU") != std::string::npos) {  // to handle GPU device variants which doesn't matter for adapter mode
+            device = "GPU";
+        }
+        auto default_mode = default_modes.find(device);
+        if(default_mode != default_modes.end()) {
+            AdapterConfig updated_config = config;
+            updated_config.set_mode(default_mode->second);
+            m_pimpl = std::make_shared<AdapterControllerImpl>(model, updated_config, prefix);
+            return;
+        } else {
+            std::string device_msg;
+            if(device.empty()) {
+                device_msg = "No device set";
+            } else {
+                device_msg = "Device \"" + device + "\" is unrecognized";
+            }
+            std::cout
+                << "[ WARNING ] " << device_msg << " to deduce default device-dependent LoRA application mode.\n"
+                << "This warning appears because no specific LoRA mode was set in AdapterConfig or MODE_AUTO was used explicitly.\n"
+                << "To avoid this warning set one of the AdapterConfig::Mode values except MODE_AUTO.";
```

Comment:
should it be exception?

Reply:
It will work with `MODE_AUTO`. Depending on use case (how dynamic your LoRA adapters are) it can fail or not, there is still chance that it work, hence it is not an exception. Anyway it is a kind of temporary stuff that we would like to eliminate when plugins become aligned.

Reply:
BTW, what if I will pass HETERO (e.g. split transformer model in SD3, FLUX to several GPUs) or AUTO ?
Is it supposed that users have to specify mode explicitly? Or we expect plugins to be aligned (even NPU and CPU?) so it's not an issue?

Reply:
If at least one GPU appears in `device` it will treat it as GPU regardless of other devices. There is a similar piece of code that treats GPU in a similar way in gen.ai, so there is no invention here. For any combination of devices that doesn't include GPU it will print this message and will leave MODE_AUTO, that is equivalent to MODE_DYNAMIC after this point. So this oversimplified device detection logic doesn't work well in case you have multiple devices in a bundle with significantly different LoRA capabilities, that's why we need to align and remove the modes except MODE_FUSE.